aptos-move/framework/aptos-framework/sources/SimpleMap.move

```diff
@@ -0,0 +1,220 @@
+/// This module provides a solution for sorted maps, that is it has the properties that
+/// 1) Keys point to Values
+/// 2) Each Key must be unique
+/// 3) A Key can be found within O(Log N) time
+/// 4) The data is stored as a sorted by Key
+/// 5) Adds and removals take O(N) time
+module AptosFramework::SimpleMap {
+    use Std::Errors;
+    use Std::Option;
+    use Std::Vector;
+    use AptosFramework::Comparator;
+
+    const EKEY_ALREADY_EXISTS: u64 = 0;
+    const EKEY_NOT_FOUND: u64 = 1;
+
+    struct SimpleMap<Key: store, Value: store> has store {
+        data: vector<Element<Key, Value>>,
+    }
+
+    struct Element<Key: store, Value: store> has store {
+        key: Key,
+        value: Value,
+    }
+
+    public fun count<Key: store, Value: store>(map: &SimpleMap<Key, Value>): u64 {
+        Vector::length(&map.data)
+    }
+
+    public fun create<Key: store, Value: store>(): SimpleMap<Key, Value> {
+        SimpleMap {
+            data: Vector::empty(),
+        }
+    }
+
+    public fun borrow<Key: store, Value: store>(
+        map: &SimpleMap<Key, Value>,
+        key: &Key,
+    ): &Value {
+        let (maybe_idx, _) = find(map, key);
+        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
+        let idx = Option::extract(&mut maybe_idx);
+        &Vector::borrow(&map.data, idx).value
+    }
+
+    public fun borrow_mut<Key: store, Value: store>(
+        map: &mut SimpleMap<Key, Value>,
+        key: &Key,
+    ): &mut Value {
+        let (maybe_idx, _) = find(map, key);
+        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
+        let idx = Option::extract(&mut maybe_idx);
+        &mut Vector::borrow_mut(&mut map.data, idx).value
+    }
+
+    public fun contains_key<Key: store, Value: store>(
+        map: &SimpleMap<Key, Value>,
+        key: &Key,
+    ): bool {
+        let (maybe_idx, _) = find(map, key);
+        Option::is_some(&maybe_idx)
+    }
+
+    public fun destroy_empty<Key: store, Value: store>(map: SimpleMap<Key, Value>) {
+        let SimpleMap { data } = map;
+        Vector::destroy_empty(data);
+    }
+
+    public fun insert<Key: store, Value: store>(
+        map: &mut SimpleMap<Key, Value>,
+        key: Key,
+        value: Value,
+    ) {
+        let (maybe_idx, maybe_placement) = find(map, &key);
```

Comment:
This increased the time of inserting compared to a simple insert in the list. I feel we should not call this map since map often indicates O(1) complexity in finding

Reply:
There are typically two types of maps LogN vs 1, here's an example: https://doc.rust-lang.org/std/collections/struct.BTreeMap.html

I just call it SimpleMap but let's figure out the right naming!

Reply:
maybe "SimpleOrderedMap", I don't have strong opinion.

Reply:
Haha, maybe, let's see where we are with data structures and names in a couple months.

Reply:
I actually had sorted in the name for other variants, but felt it added only an extra 6 characters to the name and not much more, since one still needs to read the doc block to understand the performance characteristics.